File: iosApp/Sources/CourtMarking/CapabilitySheet.swift
import Shared
import SwiftUI

/// What each of the two runs produces, opened from the "?" beside
/// "What to analyze".
///
/// Every string comes from `Shared`, so this sheet and its Android twin cannot
/// drift: the copy is a promise about what the app does, and the two phones
/// making different promises is the failure the shared analytics rules exist
/// to prevent.
///
/// Two columns rather than a paragraph. The coach is standing in a sports hall
/// deciding between two buttons, and the question is a comparison.
struct CapabilitySheet: View {
    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    Text(AnalysisCapabilityKt.CAPABILITY_SHARED_LINE)
                        .shuttlType(ShuttlType.bodyMedium)
                        .foregroundStyle(Shuttl.textTertiary)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.top, 4)

                    row(what: "", cloud: "In cloud", device: "On device", header: true)
                        .padding(.top, 20)
                    ForEach(AnalysisCapabilityKt.panelCapabilities, id: \.panel) { capability in
                        Divider().overlay(Shuttl.border)
                        row(
                            what: capability.panel.label,
                            cloud: capability.cloud,
                            device: capability.device
                        )
                    }

                    notes(title: "In cloud", lines: AnalysisCapabilityKt.cloudNotes)
                        .padding(.top, 24)
                    notes(title: "On device", lines: AnalysisCapabilityKt.deviceNotes)
                        .padding(.top, 16)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 20)
                .padding(.bottom, 32)
            }
            .background(Shuttl.bg)
            .navigationTitle("What each run produces")
            .navigationBarTitleDisplayMode(.inline)
        }
    }

    /// One line of the table.
    ///
    /// Weighted rather than fixed widths so the longest cell - "Both players,
    /// if the video is on this phone" - wraps instead of pushing the row off
    /// the screen at an accessibility text size.
    @ViewBuilder
    private func row(what: String, cloud: String, device: String, header: Bool = false) -> some View {
        HStack(alignment: .top, spacing: 12) {
            Text(what)
                .shuttlType(ShuttlType.labelMedium)
                .foregroundStyle(header ? Shuttl.textTertiary : Shuttl.text)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(cloud)
                .shuttlType(header ? ShuttlType.labelMedium : ShuttlType.bodySmall)
                .foregroundStyle(header ? Shuttl.textTertiary : Shuttl.text)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(device)
                .shuttlType(header ? ShuttlType.labelMedium : ShuttlType.bodySmall)
                .foregroundStyle(header ? Shuttl.textTertiary : Shuttl.text)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .fixedSize(horizontal: false, vertical: true)
        .padding(.vertical, 12)
    }

    @ViewBuilder
    private func notes(title: String, lines: [String]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title)
                .shuttlType(ShuttlType.labelMedium)
                .foregroundStyle(Shuttl.text)
            ForEach(lines, id: \.self) { line in
                // The bullet in its own column rather than inline in the
                // string: a wrapped line has to hang under the text, not run
                // back to the margin and read as a new bullet.
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text("\u{2022}")
                        .shuttlType(ShuttlType.bodySmall)
                        .foregroundStyle(Shuttl.textTertiary)
                    Text(line)
                        .shuttlType(ShuttlType.bodySmall)
                        .foregroundStyle(Shuttl.textTertiary)
                        .fixedSize(horizontal: false, vertical: true)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

File: androidApp/src/main/java/com/badmintontracker/android/localvideo/court/CapabilitySheet.kt
package com.badmintontracker.android.localvideo.court

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.badmintontracker.android.ui.theme.ShuttlTheme
import com.badmintontracker.shared.analytics.CAPABILITY_SHARED_LINE
import com.badmintontracker.shared.analytics.cloudNotes
import com.badmintontracker.shared.analytics.deviceNotes
import com.badmintontracker.shared.analytics.panelCapabilities

/**
 * What each of the two runs produces, opened from the "?" beside "What to
 * analyze".
 *
 * Every string comes from `:shared`, so this sheet and its iOS twin cannot
 * drift: the copy is a promise about what the app does, and the two phones
 * making different promises is the failure the shared analytics rules exist
 * to prevent.
 *
 * Two columns rather than a paragraph. The coach is standing in a sports hall
 * deciding between two buttons, and the question is a comparison.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CapabilitySheet(onDismiss: () -> Unit) {
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = MaterialTheme.colorScheme.surface,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(bottom = 32.dp),
        ) {
            Text(
                "What each run produces",
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.onBackground,
                modifier = Modifier.semantics { heading() },
            )
            Text(
                CAPABILITY_SHARED_LINE,
                style = MaterialTheme.typography.bodyMedium,
                color = ShuttlTheme.extended.textTertiary,
                modifier = Modifier.padding(top = 8.dp),
            )

            Spacer(Modifier.height(20.dp))
            CapabilityRow("", "In cloud", "On device", header = true)
            panelCapabilities.forEach { row ->
                HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
                CapabilityRow(row.panel.label, row.cloud, row.device)
            }

            Spacer(Modifier.height(24.dp))
            NoteBlock("In cloud", cloudNotes)
            Spacer(Modifier.height(16.dp))
            NoteBlock("On device", deviceNotes)
        }
    }
}

/**
 * One line of the table.
 *
 * The three columns are weighted rather than fixed so the longest cell - "Both
 * players, if the video is on this phone" - wraps instead of pushing the row
 * off the screen at a large font scale.
 */
@Composable
private fun CapabilityRow(what: String, cloud: String, device: String, header: Boolean = false) {
    val style = if (header) MaterialTheme.typography.labelMedium else MaterialTheme.typography.bodySmall
    val color = if (header) ShuttlTheme.extended.textTertiary else MaterialTheme.colorScheme.onBackground
    Row(
        modifier = Modifier.fillMaxWidth().padding(vertical = 12.dp),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(
            what,
            style = if (header) style else MaterialTheme.typography.labelMedium,
            color = if (header) color else MaterialTheme.colorScheme.onBackground,
            modifier = Modifier.weight(0.8f),
        )
        Text(cloud, style = style, color = color, modifier = Modifier.weight(1.1f))
        Text(device, style = style, color = color, modifier = Modifier.weight(1f))
    }
}

@Composable
private fun NoteBlock(title: String, notes: List<String>) {
    Column {
        Text(
            title,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onBackground,
            modifier = Modifier.semantics { heading() },
        )
        notes.forEach { note ->
            // The bullet in its own column rather than inline in the string:
            // a wrapped line has to hang under the text, not run back to the
            // margin and read as a new bullet.
            Row(modifier = Modifier.padding(top = 6.dp)) {
                Text(
                    "\u2022",
                    style = MaterialTheme.typography.bodySmall,
                    color = ShuttlTheme.extended.textTertiary,
                    modifier = Modifier.width(18.dp),
                )
                Text(
                    note,
                    style = MaterialTheme.typography.bodySmall,
                    color = ShuttlTheme.extended.textTertiary,
                )
            }
        }
    }
}
